File: elixir/1.ex
# If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.
#
# Find the sum of all the multiples of 3 or 5 below 1000.

time = :os.system_time(:micro_seconds)
Enum.reduce(0..999, fn
    a, sum when (rem(a, 3) == 0 or rem(a, 5) == 0) -> a + sum
    _, sum -> sum
    end
)
|> IO.puts
IO.puts 'time: #{ (:os.system_time(:micro_seconds) - time) / 1000000} seconds'

time = :os.system_time(:micro_seconds)
Stream.filter(0..999, &(rem(&1, 3) == 0 or rem(&1, 5) == 0))
|> Enum.sum
|> IO.puts
IO.puts 'time: #{ (:os.system_time(:micro_seconds) - time) / 1000000} seconds'

time = :os.system_time(:micro_seconds)
Enum.take_every(0..999, 3)
|> Enum.concat(Enum.take_every(0..999, 5))
|> Enum.uniq
|> Enum.sum
|>IO.puts
IO.puts 'time: #{ (:os.system_time(:micro_seconds) - time) / 1000000} seconds'

File: ruby/1.rb
# If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.
#
# Find the sum of all the multiples of 3 or 5 below 1000.cd ~/l

require 'active_support/core_ext/enumerable'
start = Time.now
sum = 0
(0..999).each do |item|
  if item % 3 == 0 || item % 5 == 0
    sum += item
  end
end

puts sum
puts Time.now - start

start = Time.now
list = (0..999).reject{ |item| (item % 3 != 0 && item % 5 != 0) }
sum = list.sum

puts sum
puts Time.now - start

start = Time.now
sum = (0..999).reduce{ |sum, item| (item % 3 == 0 || item % 5 == 0) ? sum + item : sum }

puts sum
puts Time.now - start

start = Time.now
sum = 0
(0..999).step(3) { |item| sum += item }
(0..999).step(5) { |item| sum += item }
(0..999).step(15) { |item| sum -= item }

puts sum
puts Time.now - start